### PR TITLE
build: include commit hash & version

### LINKED ISSUE #2304: Include commit hash or tag in version string
This could be done in way similar to Semver, string after `+` is additional version information that is not used for version comparison. 

To get hash or tag from git cjdns uses:

```
git describe --always --dirty --tags
```

This would allow precise identification of gateway, and APIs during development periods.
(I wouldn't have to ask Larsg if given patch was deployed or not)

### PR DESCRIPTION
**Pull request by hutenosa**
(partially) fixes #2304 

New script `bin/genversion` generates SemVer-compatible string out of closest git tag. The script is called in Makefile to pass IPFS version and commit info as linker arguments.
Dockerfiles got optional build argument `IPFS_VERSION` to avoid including `.git/` into the image. 

License: MIT
Signed-off-by: hutenosa <hutenosa@mm.st>

### CODE REVIEW

**Comment by Kubuxu** [comment]
IIRC there was API version comparison somewhere, IMO it should work among the same pre-release, different commit metadata. You can try it by building, starting daemon, adding commit, building, trying to use `ipfs files ls /`.

**Comment by hutenosa** [comment]
Travis uses quite outdated docker (1.8.2) which doesn't support build arguments (appeared in 1.9.0). I was unable to upgrade docker in Travis environment so I've just replaced `ARG` with `ENV` in both Dockerfiles. Drawback: docker builds will all have hardcoded version `0.4.0-dev` unless explicitly changed. 

Once Travis refreshes its docker (they promised to keep it updated), we can go back to build arguments and pass the version via cmdline: `docker build --build-arg IPFS_VERSION=...`

@Kubuxu I checked, client and server versions should match and it's about version tags like `0.4.0-rc2`. Commit metadata can be different. Since now version is SemVer-compatible string, we may soften the condition and compare only major versions starting from `1.0.0`

**Comment by Kubuxu** [comment]
What about this: https://github.com/hutenosa/go-ipfs/blob/fix/version-commit-hash/commands/http/handler.go#L453-L463

**Comment by hutenosa** [comment]
`config.ApiVersion` contains only the version number and not the current commit: https://github.com/hutenosa/go-ipfs/blob/fix/version-commit-hash/repo/config/version.go#L18

However, client and daemon with different version number / pre-release will not be able to communicate.

**Comment by Kubuxu** [comment]
Right, sorry for trouble.

Go/Bash LGTM, but I can't say anything about docker.

**Comment by hutenosa** [comment]
@lgierth Are you fine with Dockefile changes?

**Comment by whyrusleeping** [comment]
@hutenosa could you rebase these changes on the lastest master? I'll also poke @lgierth to have him take a look at the dockerfile changes.

File: repo/config/version.go [hunk 1]
```diff
@@ -7,13 +7,15 @@ import (
...
 )
 
+// CurrentVersionNumber and CurrentCommit are normally set via ldflags -X
+
 // CurrentCommit is the current git commit, this is set as a ldflag in the Makefile
 var CurrentCommit string
 
 // CurrentVersionNumber is the current application's version literal
-const CurrentVersionNumber = "0.4.0-dev"
+var CurrentVersionNumber = "0.0.0-dev"
```
Comment (lgierth) [inline repo/config/version.go]: `CurrentVersionNumber` should stay as it is and not carry commit information -- it's used in too many places where the assumption is a simple vN.N.N string. Maybe just use CurrentCommit?

File: repo/config/version.go [hunk 2]
```diff
@@ -7,13 +7,15 @@ import (
...
 
-const ApiVersion = "/go-ipfs/" + CurrentVersionNumber + "/"
+var ApiVersion = "/go-ipfs/" + CurrentVersionNumber + "/"
```
Comment (lgierth) [inline repo/config/version.go]: Does this do what expected when changing CurrentVersionNumber via ldflags? In any case the same vN.N.N expectation applies here.

**Comment by lgierth** [comment]
Hey @hutenosa I'm so sorry for letting this sit so long. I like this PR very much!

I think this can be simplified a bit, by not touching CurrentVersionNumber,
and instead introducing a "full version" as well as deprecating "current commit".

- [ ] leave CurrentVersionNumber (and ApiVersion) untouched
- [ ] FullVersion, passed as IPFS_FULL_VERSION
- [ ] add FullVersion to `/api/v0/version`
- [ ] add `ipfs version --full` option
- [ ] add Full Version to `:8080/version`
- [ ] deprecate/remove mentions of CurrentCommit
  - deprecate `ipfs version --commit`
  - deprecate Commit in `/api/v0/version`
  - remove Commit in `:8080/version`

I can help you out with a few of these any time. 

Regarding git, I haven't found a way to use git describe without a full repo,
so passing the full version via docker is okay.

**Comment by hutenosa** [comment]
I'm closing this PR since I deleted the forked repository. Unfortunately, github has no means to rebind the PR to a new fork. I'll resubmit a new one shortly.

**Pull request closed without merging**